src/bobtail/dag.cpp

```diff
@@ -117,6 +86,82 @@ void CBobtailDag::UpdateCompatibility(const int16_t &new_id, const std::set<int1
     }
 }
 
+void CBobtailDag::UpdateDagScore()
+{
+    // keep track of what has been mapNodeScore
+    std::map<CDagNode*, uint64_t> mapNodeScore;
+    // build out the dag by level, a nodes level is determined by
+    // its shortest path to a base
+    std::vector<std::set<CDagNode*> > leveled_dag;
+    // first find the bases
+    leveled_dag.emplace_back();
+    bool do_another_level = false;
+    // there is always at least 1 node in a dag
+    for (auto &node : _dag)
+    {
+        if (node->IsBase())
+        {
+            leveled_dag[0].emplace(node);
+            mapNodeScore.emplace(node, 1);
+            if (node->descendants.empty() == false)
+            {
+                do_another_level = true;
+            }
+        }
+    }
+    while (do_another_level == true)
+    {
+        size_t search_index = leveled_dag.size() - 1;
+        // add another level
+        leveled_dag.emplace_back();
+        do_another_level = false;
+        for (auto &node : leveled_dag[search_index])
+        {
+            for (auto &desc : node->descendants)
+            {
+                if (mapNodeScore.count(desc) == 0)
+                {
+                    mapNodeScore.emplace(desc, 1);
+                    leveled_dag.back().emplace(desc);
+                    if (desc->IsTip() == false)
```

Comment:
What's the difference between `desc->IsTip() == false` used here and `node->descendants.empty() == false` used above?